src/bin/cargo-flamegraph.rs

```diff
@@ -256,116 +293,59 @@ fn build(opt: &Opt) {
-                    let name = file_name.to_string_lossy();
-                    name.starts_with(bin)
-                        && !name.ends_with(".d")
-                } else {
-                    false
-                }
-            })
-            .and_then(|r| r.ok())
-        })
-        .and_then(|f| {
-            f.path()
-                .file_name()
-                .map(|n| n.to_string_lossy().to_string())
-        })
-        .unwrap_or_else(|| {
-            eprintln!(
-                "could not find desired target {} \
-                 in the {} targets for this crate",
-                bin, ty
-            );
-            std::process::exit(1);
-        })
+    artifacts
 }
 
-fn workload(opt: &Opt) -> Vec<String> {
-    let mut metadata_cmd =
-        cargo_metadata::MetadataCommand::new();
-    metadata_cmd.no_deps();
-    let metadata = metadata_cmd
-        .exec()
-        .expect("could not access crate metadata");
-
-    let mut binary_path = metadata.target_directory;
-
-    if let Ok(t) = std::env::var("CARGO_BUILD_TARGET") {
-        binary_path.push(t);
-    }
-
-    if opt.dev {
-        binary_path.push("debug");
-    } else {
-        binary_path.push("release");
-    }
-
-    if opt.example.is_some() {
-        binary_path.push("examples");
-    } else if opt.bench.is_some() {
-        binary_path.push("deps");
-    }
-
-    let targets: Vec<String> = metadata
-        .packages
-        .into_iter()
-        .flat_map(|p| p.targets)
-        .filter(|t| t.crate_types.contains(&"bin".into()))
-        .map(|t| t.name)
-        .collect();
-
-    if targets.is_empty() {
-        eprintln!("no Rust binary targets found");
-        std::process::exit(1);
-    }
+fn workload(
+    opt: &Opt,
+    artifacts: &[Artifact],
+) -> Vec<String> {
+    let binary_path = {
+        let target = opt.target_name();
+        let kind = opt.target_kind().to_owned();
+
+        if artifacts.iter().all(|a| a.executable.is_none())
+        {
+            eprintln!( "build artifacts do not contain any executable to profile");
+            std::process::exit(1);
+        }
 
-    let target = if let Some(ref test) = opt.test {
-        find_binary("test", &binary_path, test)
-    } else if let Some(ref bench) = opt.bench {
-        find_binary("bench", &binary_path, bench)
-    } else if let Some(ref bin) =
-        opt.bin.as_ref().or_else(|| opt.example.as_ref())
-    {
-        if targets.contains(&bin) {
-            bin.to_string()
-        } else {
+        // target.kind is an array for some reason. No idea why though, it always seems to contain exactly one element.
+        // If you know why, feel free to PR and handle kind properly.
+        let artifact = artifacts.iter().find(|a| {
+            a.target.name == target
+                && a.target.kind.contains(&kind)
+                && a.executable.is_some()
+        });
+
+        let artifact = artifact.unwrap_or_else(|| {
```

Comment:
Let's chain this directly? I think we can return `artifacts.iter().find(...).unwrap_or_else().executable.as_deref().expect()`? Also I'd probably leave that outer `expect()` to the caller.

Reply:
> I think we can return ...?

In this commit, yes but in the next commit 010ad1a8a6d3b04dce6dc49ada8e1641d669e282 I'm adding the debug info check inbetween `unwrap_or_else` and `executable`. This check needs the artifact.

> Also I'd probably leave that outer expect() to the caller.

It's not a function anymore just a block expression. Also the expect itself should never ever occur unless someone changes this part of the code again because the `find` predicate only allows artifacts with executables.
```rust
let binary_path = {
    let artifact = artifacts.iter().find(...).unwrap_or_else(...);
    artifact.executable.as_deref().expect(...)
};
```

Reply:
Can we express that in some way that we don't have the unnecessary `expect()` call, then? I guess we could `find_map()` in a way that we return the deref of executable directly along with the `artifact`?

Reply:
Excellent suggestion. I like the new version much more.